neqo-transport/tests/server.rs

```diff
@@ -900,40 +900,56 @@ fn has_active_connections() {
 #[test]
 fn saved_datagrams() {
     let mut server = default_server();
-    let mut client = default_client();
 
-    // Any packet will do, but let's make something that looks real.
-    let damaged_dgram = {
-        let dgram = client.process_output(now()).dgram().expect("a datagram");
-        let mut input = dgram.to_vec();
-        input[1] ^= 0x12;
-        Datagram::new(
-            dgram.source(),
-            dgram.destination(),
-            dgram.tos(),
-            input.clone(),
-        )
+    let valid_dgram = {
+        let mut client = default_client();
+        client.process_output(now()).dgram().expect("a datagram")
     };
 
-    // Server sends a version negotation immediately. Saves second input datagram for later.
+    // Any packet will do, but let's make something that looks real.
+    let mut invalid_dgrams: Vec<_> = [0x12, 0x13, 0x14]
+        .into_iter()
+        .map(|damage| {
+            let mut client = default_client();
+            let dgram = client.process_output(now()).dgram().expect("a datagram");
+            let mut input = dgram.to_vec();
+            input[1] ^= damage;
+            Datagram::new(
+                dgram.source(),
+                dgram.destination(),
+                dgram.tos(),
+                input.clone(),
+            )
+        })
+        .collect();
+
+    // Server sends a version negotation immediately. Saves second and third
+    // input datagram for later.
     server
         .process_multiple(
-            vec![damaged_dgram.clone(), damaged_dgram.clone()],
+            vec![
+                invalid_dgrams.pop().unwrap(),
```

Comment:
Rather than collect datagrams into a vec and then pop them, why not make an `invalid_dgram()` function, passing it the damage you want it to inflict.

You could even avoid making a new client, instead damaging `valid_dgram`, though you would have to clone it.

Reply:
> Rather than collect datagrams into a vec and then pop them, why not make an invalid_dgram() function, passing it the damage you want it to inflict.

:+1: will give it a try. Thanks.

> You could even avoid making a new client, instead damaging valid_dgram, though you would have to clone it.

I believe I need different connection IDs to force the server to allocate new connections.